### DISCUSSION TITLE
Problem building Native windows app

### DISCUSSION THREAD

**Discussion by markdemich** [airsdk/Adobe-Runtime-Support#2479] [Q&A] [answered]
I have a very large app I'm trying to move from Flash Builder into VS Code using the vscode extension by Bowler Hat LLC.  To start, I created a new project with a simple hello world from their documentation.  I'm trying to package with with target type "native".  It is building an EXE.  However, when I try to run it, it prompt me to install the air runtime.

I even tried running ADT from the command prompt, but the resultant exe still asks me to download the runtime.

```
adt -package -storetype pkcs12 -keystore ../cert/xxxx.pfx -target native  main.exe  Main-app.xml Main.swf
```
I'm not exactly sure what is going on.

ADT is version 50.1.1.2

Also, if I tried to put another -storetype and -keystore after -target native in order to sign the final main.exe, it gives me an error saying it can't use the certificate file.  I suspect I'm missing something simple.   Any help would be appreciated.

**Comment by FliplineStudios** [comment, marked as answer]
> I even tried running ADT from the command prompt, but the resultant exe still asks me to download the runtime.

I'm pretty sure the older "native" target just makes an EXE installer for a classic-style AIR application, which requires the user to have the AIR Runtime installed as separate software.  If you want something standalone, I would create a captive runtime bundle using the "bundle" target instead of native.  I'm not sure why neither the Adobe nor the Harman documentation seem to mention it within the adt package commands, but you can find info about it and the syntax here:

https://help.adobe.com/en_US/air/build/WSfffb011ac560372f709e16db131e43659b9-8000.html

**Reply by markdemich** [reply]
@FliplineStudios , Thanks for the quick reply.   Do you know if there is an option that will automatically download the runtime?  Or is that gone too?  I would really hate to have to write an installer for this.

Once I hear back from you, I will mark this as answered.  Thanks again.

**Reply by FliplineStudios** [reply]
I'm not really sure, I thought the "native" installers were supposed to automatically download the latest runtime but I haven't used them myself (I've only used "bundle" since they include their own runtime).  At least as of AIR 33.1.1.795 it sounds like Harman updated the native installers to point to the AIR Runtime downloads from Harman instead of Adobe, I wonder if there's some issue happening with that...

Is this for supporting an existing app that's already relying on having the separate AIR runtime instead of being bundled with it?

**Reply by markdemich** [reply]
From what I understand, the app needs native so it can support some features that otherwise would not work.  My hope was to include the entire runtime.  That worked with air 32.  I have to deploy to both Windows and Mac in a seamless way.  Thank for your input.

**Reply by ajwfrost** [reply]
Yes just to confirm what @FliplineStudios has said - the "bundle" option will include the runtime, the "native" option creates an installer - this doesn't have the ability to download the runtimes if they're not there though (or rather, we don't have the server-side infrastructure set up to deliver this) so if you've not already got an appropriate AIR shared runtime installed, the "native" option is no good.

On macOS it's fairly trivial to create a bundle and from that to create a .pkg installer file. On Windows it's more of a challenge as the tools required for this aren't available on everyone's computers I believe.. but we were going to investigate the idea of including the runtime in with that native package. There may be an issue with signing some of these packages, in the way they're currently done, because of improvements in security combined with the fact that again we don't have the "proper" signing tools always available on Windows.

I'll ask someone to investigate this; thinking about it, it might be that having the .exe format pull down our latest runtime is the simplest option to get back the earlier capabilities...

thanks

**Reply by markdemich** [reply]
@ajwfrost , what is your recommendation for me.  I need a way for users to click on a download link from a website for both MacOS and Windows and make it easy for them to install my air application in the easiest way possible.  My app also needs access to some native abilities.  I'm afraid if I use bundle, I won't get all the native support.  Is that true?  I have a tool INNO Setup that I can create a Windows installer.  Should I just set up that installer to manually bundle the runtime and install it first, then install my native package on Windows?  Or should I just use the bundle target and do the same?  Any advice would be appreciated.

**Reply by ajwfrost** [reply]
Hi -> the "bundle" approach is best here, it does give you access to all the native capabilities (ANEs, NativeProcess etc); it's only when you package an application as a .air file that you don't get those. So, if you use ADT to create a "bundle", what it will do is create a folder that contains the redistributable files in the appropriate structure. If you can then use the INNO Setup program to turn those into an installer package, this is then something that you can distribute and that should work on a machine regardless of whether it's got the AIR shared runtime already installed or not (and independently of that, in case someone had an old version of it..)

thanks
